internal/pn_counter.go

```diff
@@ -0,0 +1,253 @@
+	maxConfiguredReplicaCount   int32
+	emptyAddresses              map[IAddress]struct{}
+	observedClock               unsafe.Pointer
+}
+
+func newPNCounterProxy(client *HazelcastClient, serviceName *string, name *string) (*PNCounterProxy, error) {
+	c := &PNCounterProxy{
+		proxy:          &proxy{client, serviceName, name},
+		emptyAddresses: make(map[IAddress]struct{}),
+	}
+	atomic.StorePointer(&c.observedClock, unsafe.Pointer(NewVectorClock()))
+	return c, nil
+}
+
+func (c *PNCounterProxy) Get() (currentValue int64, err error) {
+	target, err := c.getCRDTOperationTarget(c.emptyAddresses)
+	if err != nil {
+		return
+	}
+	if target == nil {
+		err = NewHazelcastNoDataMemberInClusterError("cannot invoke operations on a CRDT because the cluster does not contain any data members", nil)
+		return
+	}
+	response, err := c.invokeGetInternal(c.emptyAddresses, nil, target)
+	if err != nil {
+		return
+	}
+	value, replicaTimestamps, _ := PNCounterGetDecodeResponse(response)()
+	c.updateObservedReplicaTimestamps(replicaTimestamps)
+	return value, nil
+}
+
+func (c *PNCounterProxy) operation(delta int64, getBeforeUpdate bool) (value int64, err error) {
+	target, err := c.getCRDTOperationTarget(c.emptyAddresses)
+	if err != nil {
+		return
+	}
+	if target == nil {
+		err = NewHazelcastNoDataMemberInClusterError("cannot invoke operations on a CRDT because the cluster does not contain any data members", nil)
+		return
+	}
+	response, err := c.invokeAddInternal(delta, getBeforeUpdate, c.emptyAddresses, nil, target)
+	if err != nil {
+		return
+	}
+	value, replicaTimestamps, _ := PNCounterAddDecodeResponse(response)()
+	c.updateObservedReplicaTimestamps(replicaTimestamps)
+	return value, nil
+}
+
+func (c *PNCounterProxy) GetAndAdd(delta int64) (previousValue int64, err error) {
+	return c.operation(delta, true)
+}
+
+func (c *PNCounterProxy) AddAndGet(delta int64) (updatedValue int64, err error) {
+	return c.operation(delta, false)
+}
+
+func (c *PNCounterProxy) GetAndSubtract(delta int64) (previousValue int64, err error) {
+	return c.operation(-delta, true)
+}
+
+func (c *PNCounterProxy) SubtractAndGet(delta int64) (updatedValue int64, err error) {
+	return c.operation(-delta, false)
+}
+
+func (c *PNCounterProxy) DecrementAndGet() (updatedValue int64, err error) {
+	return c.operation(-1, false)
+}
+
+func (c *PNCounterProxy) IncrementAndGet() (updatedValue int64, err error) {
+	return c.operation(1, false)
+}
+
+func (c *PNCounterProxy) GetAndDecrement() (previousValue int64, err error) {
+	return c.operation(-1, true)
+}
+
+func (c *PNCounterProxy) GetAndIncrement() (previousValue int64, err error) {
+	return c.operation(1, true)
+}
+
+func (c *PNCounterProxy) Reset() {
+	atomic.StorePointer(&c.observedClock, unsafe.Pointer(NewVectorClock()))
+}
+
+func (c *PNCounterProxy) getCRDTOperationTarget(excludedAddresses map[IAddress]struct{}) (IAddress, error) {
+	c.targetSelectionMutex.RLock()
+	defer c.targetSelectionMutex.RUnlock()
```

Comment:
You can just store whatever is read into a local variable before releasing the lock. Then release the lock and return the stored value. `defer` is not unavoidable.

Reply:
Why do we try to avoid `defer`? Also, I think storing to a local variable is reverse to logic of the concurrency. After releasing lock, what if other goroutine writes and changes the value. We won't be able to get the new value of shared variable.